### ISSUE TITLE
How to change parameters of models during training based on their performance?

### ISSUE THREAD

**Issue by btseytlin** [KevinMusgrave/powerful-benchmarker#62]
I have implemented a weighted sampler: it selects more examples of some classes, based on provided weights.

Every few epochs, I want to change weights based on per-class f1-score on the training set.

For experiments I managed to do this with `pytorch-metric-learning` by adding some ugly code to testing hooks. Now I want to benchmark the performance of this method, so I want to implement this method in `powerful-benchmarker`.

So I need to do the following every few epochs:
* Run KNN on the train set, obtain f1-scores
* Update a parameter of a custom sampler

How can I do it? Currently I see no way to add a "hook" to training. Should I subclass a trainer to implement that logic?

**Comment by KevinMusgrave** [comment]
You can extend the existing [HookFactory](https://github.com/KevinMusgrave/powerful-benchmarker/blob/master/src/powerful_benchmarker/factories/hook_factory.py) class, register it, and then specify it in your script like this: ```--factories {hook: {YourHookFactory: {}}}```

When you extend that class, you probably just need to override ```create_end_of_epoch_hook```. You can access the epoch and sampler via ```trainer.epoch``` and ```trainer.sampler```.

**Comment by btseytlin** [comment]
Great, thanks

**Comment by btseytlin** [comment]
FYI I had to use this to get it to work:
```
--factories {hook~OVERRIDE~: {CustomHookFactory: {}}}
```
